File: validation/mml_pcm_quadrature_diagnosis.py
#!/usr/bin/env python3
"""Separate finite-GH optimization and integration error on one observed PCM fixture."""
import argparse
from dataclasses import replace
import json
from pathlib import Path
import platform
import subprocess
import sys

import numpy as np
import pandas as pd
import scipy

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))
import streamlit_app as app
from mfrm_app.mml_engine_v2 import run_free_sd_stationarity_v2
from validation.mml_python_analytic_sd_probe import native_fit, scores, sha256, OPTIONS
from validation.mml_free_sd_cross_language_probe import dump, difference
from validation.mml_free_sd_stationarity_adapter import prepare_app_free_sd_problem
from validation.mml_free_sd_stationarity_known_fixture import _json_value

PREVIOUS = ROOT / "validation/mml_python_analytic_sd_20260912"
SUMMARY_SHA256 = "02315c5b9f10c9a9ebd475be25bee81028632dcec1d46a14fb83c0ef8b65e3f2"
START_ORDERS = (31, 61, 121)
ORDERS = (*START_ORDERS, 181)
# Implementation/replay checks, fixed before this run; no scientific margins.
LIMITS = {"replay": 1e-8, "r_nll": 1e-9, "r_score": 1e-10,
          "r_probability": 1e-12, "continuous_refinement": 1e-8,
          "relative_mass_error": 1e-8, "gradient": 1e-4}
INTEGRATION = [
    {"bound": 12, "rel_tol": 1e-10, "abs_tol": 1e-12},
    {"bound": 14, "rel_tol": 1e-12, "abs_tol": 1e-13},
]


def evaluate(base, point):
    expanded = app.expand_params(point[:-1], base.sizes, base.config)
    # The independent R code reconstructs these same coordinates, not these arrays.
    eta = app.compute_base_eta(base.idx, expanded, base.config)
    cumulative = np.column_stack([np.zeros(5), np.cumsum(expanded["steps_mat"], axis=1)])
    probabilities = [app.category_prob_pcm(eta + theta, cumulative, base.idx["step_idx"])[:20]
                     for theta in (-3, 0, 3)]
    return {"coordinates": point.tolist(), "sigma": float(np.exp(point[-1])),
            "finite": {str(q): scores(replace(base, quadrature_points=q), point) for q in ORDERS},
            "probabilities": np.asarray(probabilities).tolist()}


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--output-dir", type=Path, required=True)
    out = parser.parse_args().output_dir.resolve()
    out.mkdir(parents=True, exist_ok=False)
    assert sha256(PREVIOUS / "summary.json") == SUMMARY_SHA256
    previous = json.loads((PREVIOUS / "summary.json").read_text())
    assert not previous["scientific_inference_ready"] and not previous["qualification_eligible"]
    for name, expected in previous["artifact_sha256"].items():
        assert sha256(PREVIOUS / name) == expected, name
    spec = json.loads((PREVIOUS / "input.json").read_text())["datasets"]["PCM"]
    data = pd.DataFrame(spec["data"])
    assert spec["seed"] == 2026091202 and len(data) == 480 and data.Person.nunique() == 24
    paths = set(previous["source_sha256"]) | {
        str(Path(__file__).relative_to(ROOT)),
        str(Path(__file__).with_suffix(".R").relative_to(ROOT)),
        "validation/mml_python_analytic_sd_probe.py",
        "validation/mml_free_sd_cross_language_probe.py",
    }
    hashes = {p: sha256(ROOT / p) for p in sorted(paths)}
    starts = {q: np.asarray(json.loads((PREVIOUS / f"PCM_q{q}_analytic.json").read_text())
                           ["restart_polish"]["joint_coordinates"]) for q in START_ORDERS}
    dump(out / "input.json", {"classification": "OBSERVED_DEVELOPMENT_ONLY",
        "scientific_inference_ready": False, "qualification_eligible": False,
        "source_sha256": hashes, "previous_summary_sha256": SUMMARY_SHA256,
        "previous_artifact_sha256": previous["artifact_sha256"],
        "data": spec["data"], "seed": spec["seed"], "forced_extreme_persons": spec["forced_extreme_persons"],
        "starts": {str(q): p.tolist() for q, p in starts.items()}, "orders": ORDERS,
        "options": OPTIONS.__dict__, "limits": LIMITS, "integration": INTEGRATION,
        "design": "3 retained starts x 4 finite GH orders; one polish and one restart per cell; no retries",
        "continuous_scope": "Same-point independent R log-PCM integrals; no continuous-objective refitting"})

    native = native_fit(data, "PCM")  # Prepare the app's existing identified design.
    base = prepare_app_free_sd_problem(native)
    assert native["prep"]["n_obs"] == 480 and not bool(native["summary"].iloc[0]["InferenceReady"])
    assert base.config["facet_names"] == ["Rater", "Criterion"]
    assert base.config["facet_levels"] == {"Rater": ["R1", "R2", "R3", "R4"],
                                           "Criterion": ["C1", "C2", "C3", "C4", "C5"]}
    assert [(k, v) for k, v in base.sizes.items() if v] == [("Rater", 3), ("Criterion", 5), ("steps", 15)]
    assert np.all(base.idx["weight"] == 1) and base.constraint_residual(starts[31][:-1]) < 1e-12
    assert list(native["prep"]["levels"]["Person"]) == [f"P{i:02d}" for i in range(1, 25)]
    cases, runs, replay = {}, {}, []
    for q, point in starts.items():
        case = evaluate(base, point)
        case.update(kind="retained", q=q)
        cases[f"retained_q{q}"] = case
        old = next(f for f in previous["fits"] if f["model"] == "PCM" and f["q"] == q)
        differences = {k: difference(case["finite"][str(q)][k], old["at_analytic_fit"][k])
                       for k in ("nll", "eap", "sd")}
        replay.append({"q": q, "differences": differences,
                       "pass": max(differences.values()) < LIMITS["replay"]})
    dump(out / "replay.json", replay)
    assert all(r["pass"] for r in replay), "Retained-point replay failed"
    print("Retained-point replay: 3/3", flush=True)

    for q in ORDERS:
        problem = replace(base, quadrature_points=q)
        for source_q, start in starts.items():
            name = f"q{q}_from_q{source_q}"
            run = run_free_sd_stationarity_v2(
                start[:-1], np.exp(start[-1]), problem.value, problem.value_gradient,
                observations=problem.observations, structural_bounds=problem.structural_bounds,
                sigma_bounds=problem.sigma_bounds, options=OPTIONS,
                constraint_residual_function=problem.constraint_residual,
                joint_value_gradient=problem.joint_value_gradient,
            )
            record = _json_value(run.to_dict())
            dump(out / f"{name}.json", record)
            runs[name] = record
            case = evaluate(base, np.asarray(run.restart_polish.joint_coordinates))
            case.update(kind="cross_start", q=q, source_q=source_q)
            cases[name] = case
            print(name, "sigma", run.final_sigma, "nll", run.final_objective,
                  "terminated", run.algorithm_terminated, "score", run.final_projected_gradient_supnorm, flush=True)
    dump(out / "python.json", cases)
    subprocess.run(["Rscript", str(Path(__file__).with_suffix(".R")),
                    str(out / "input.json"), str(out / "python.json"), str(out)], check=True)

    comparisons = []
    for name, case in cases.items():
        r = json.loads((out / f"{name}_r.json").read_text())
        same = {str(q): {k: difference(case["finite"][str(q)][k], r["finite"][str(q)][k])
                        for k in ("nll", "eap", "sd")} for q in ORDERS}
        probability = difference(case["probabilities"], r["probabilities"])
        primary, tight = r["continuous"]
        refinement = {k: difference(primary[k], tight[k]) for k in ("nll", "eap", "sd")}
        integration_ok = all(
            c["numeric_relative_mass_error_sum"] < LIMITS["relative_mass_error"] and
            c["tail_relative_mass_bound_sum"] < LIMITS["relative_mass_error"]
            for c in (primary, tight))
        errors = {str(q): {k: difference(case["finite"][str(q)][k], tight[k])
                           for k in ("nll", "eap", "sd")} for q in ORDERS}
        passed = (probability < LIMITS["r_probability"] and integration_ok and
                  max(refinement.values()) < LIMITS["continuous_refinement"] and
                  all(d["nll"] < LIMITS["r_nll"] and max(d["eap"], d["sd"]) < LIMITS["r_score"]
                      for d in same.values()))
        comparisons.append({"id": name, "kind": case["kind"], "q": case["q"],
            "sigma": case["sigma"], "same_point_r_differences": same,
            "probability_difference": probability, "continuous_refinement": refinement,
            "continuous_nll": tight["nll"], "continuous_log_sigma_nll_score": tight["log_sigma_nll_score"],
            "continuous_errors": errors, "integration_error_checks_pass": integration_ok,
            "pass": passed})
    spread = []
    for q in ORDERS:
        names = [f"q{q}_from_q{s}" for s in START_ORDERS]
        group = [cases[n] for n in names]
        spread.append({"q": q, "ids": names,
            "sigma_range": [min(c["sigma"] for c in group), max(c["sigma"] for c in group)],
            "finite_nll_range": [min(c["finite"][str(q)]["nll"] for c in group),
                                 max(c["finite"][str(q)]["nll"] for c in group)],
            "max_coordinate_distance": max(difference(a["coordinates"], b["coordinates"])
                                           for a in group for b in group)})
    terminated = all(r["algorithm_terminated"] for r in runs.values())
    stationary = all(r["final_projected_gradient_supnorm"] < LIMITS["gradient"] for r in runs.values())
    assert hashes == {p: sha256(ROOT / p) for p in hashes}, "Sources changed during run"
    assert sha256(PREVIOUS / "summary.json") == SUMMARY_SHA256
    assert all(sha256(PREVIOUS / n) == h for n, h in previous["artifact_sha256"].items())
    passed = all(c["pass"] for c in comparisons) and terminated and stationary
    dump(out / "summary.json", {"classification": "OBSERVED_DEVELOPMENT_ONLY",
        "scientific_inference_ready": False, "qualification_eligible": False,
        "implementation_checks_pass": passed, "same_point_checks_pass": all(c["pass"] for c in comparisons),
        "all_refits_terminated": terminated, "all_refits_small_projected_gradient": stationary,
        "comparisons": comparisons, "cross_start_spread": spread,
        "limits": LIMITS, "source_sha256": hashes,
        "runtime": {"python": platform.python_version(), "numpy": np.__version__,
                    "scipy": scipy.__version__, "platform": platform.platform()},
        "artifact_sha256": {p.name: sha256(p) for p in sorted(out.iterdir()) if p.is_file()},
        "limitations": ["One observed stress dataset; not a recovery or qualification study",
            "Adaptive integral error estimates are not rigorous interval certificates",
            "Three observed starts do not prove global optimality",
            "No continuous-objective optimum, SE, CI or coverage qualification",
            "All primary and restart termination codes are retained, including failures"]})
    print("Implementation checks:", passed, "scientific inference ready: False", flush=True)
    return 0 if passed else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: validation/mml_pcm_quadrature_diagnosis.R
#!/usr/bin/env Rscript
# Independent scalar PCM evaluation for the fixed 24 x 4 x 5, score 0:4 fixture.
# Adapt the earlier mode-split R integration to this design, preserving old files.
args <- commandArgs(TRUE)
stopifnot(length(args) == 3L)
input <- jsonlite::read_json(args[1], simplifyVector = TRUE)
cases <- jsonlite::read_json(args[2])
stopifnot(identical(input$classification, "OBSERVED_DEVELOPMENT_ONLY"),
          !input$scientific_inference_ready, !input$qualification_eligible)
data <- input$data
stopifnot(nrow(data) == 480L, all(data$Score %in% 0:4),
          identical(sort(unique(data$Person)), sprintf("P%02d", 1:24)),
          all(data$Rater %in% paste0("R", 1:4)), all(data$Criterion %in% paste0("C", 1:5)),
          all(data$Score[data$Person == "P01"] == 0),
          all(data$Score[data$Person == "P02"] == 4))
person <- match(data$Person, sprintf("P%02d", 1:24))
rater <- match(data$Rater, paste0("R", 1:4))
criterion <- match(data$Criterion, paste0("C", 1:5))
rows_by_person <- lapply(1:24, function(p) which(person == p))
stopifnot(all(vapply(rows_by_person, function(r) length(r) == 20L &&
  nrow(unique(data[r, c("Rater", "Criterion")])) == 20L, logical(1))))
rules <- lapply(input$orders, statmod::gauss.quad.prob, dist = "normal")
names(rules) <- as.character(input$orders)
stopifnot(all(vapply(rules, function(r) all(is.finite(r$nodes)) && all(r$weights > 0) &&
  abs(sum(r$weights) - 1) < 1e-12 && abs(sum(r$weights * r$nodes^2) - 1) < 1e-12,
  logical(1))))

expand_parameters <- function(par) {
  stopifnot(length(par) == 24L, all(is.finite(par)))
  free_steps <- matrix(par[9:23], nrow = 5, byrow = TRUE)
  steps <- cbind(free_steps, -rowSums(free_steps))
  list(rater = c(par[1:3], -sum(par[1:3])), criterion = par[4:8],
       cumulative = t(apply(steps, 1, function(s) c(0, cumsum(s)))), sigma = exp(par[24]))
}
log_probabilities <- function(x, theta, rows = seq_len(nrow(data))) {
  base <- theta - x$rater[rater[rows]] - x$criterion[criterion[rows]]
  logits <- outer(base, 0:4) - x$cumulative[criterion[rows], , drop = FALSE]
  shifted <- logits - apply(logits, 1, max)
  shifted - log(rowSums(exp(shifted)))
}
finite_summary <- function(x, rule) {
  theta <- x$sigma * rule$nodes
  log_likelihood <- vapply(theta, function(t) {
    p <- log_probabilities(x, t)
    selected <- p[cbind(seq_len(nrow(data)), data$Score + 1L)]
    vapply(rows_by_person, function(rows) sum(selected[rows]), numeric(1))
  }, numeric(24))
  joint <- sweep(log_likelihood, 2, log(rule$weights), "+")
  center <- apply(joint, 1, max)
  mass <- rowSums(exp(joint - center))
  posterior <- exp(joint - center) / mass
  eap <- as.numeric(posterior %*% theta)
  deviations <- matrix(theta, 24, length(theta), byrow = TRUE) - eap
  list(nll = -sum(center + log(mass)), eap = eap,
       sd = sqrt(rowSums(posterior * deviations^2)))
}
continuous_summary <- function(x, settings) {
  bound <- settings$bound
  per_person <- lapply(rows_by_person, function(rows) {
    log_integrand <- function(z) vapply(z, function(v) {
      p <- log_probabilities(x, x$sigma * v, rows)
      sum(p[cbind(seq_along(rows), data$Score[rows] + 1L)]) + dnorm(v, log = TRUE)
    }, numeric(1))
    # Log PCM likelihood is concave in theta; adding log phi(z) makes it strictly concave.
    mode <- optimize(log_integrand, c(-bound, bound), maximum = TRUE, tol = 1e-10)
    stopifnot(abs(mode$maximum) < bound - 0.1)
    cuts <- sort(unique(c(-bound, 0, mode$maximum, bound)))
    integrals <- lapply(0:2, function(k) {
      f <- function(z) exp(log_integrand(z) - mode$objective) * z^k
      parts <- lapply(seq_len(length(cuts) - 1L), function(i) integrate(
        f, cuts[i], cuts[i + 1L], subdivisions = 500L,
        rel.tol = settings$rel_tol, abs.tol = settings$abs_tol))
      stopifnot(all(vapply(parts, function(p) identical(p$message, "OK"), logical(1))))
      c(value = sum(vapply(parts, `[[`, numeric(1), "value")),
        error = sum(vapply(parts, `[[`, numeric(1), "abs.error")))
    })
    values <- vapply(integrals, `[[`, numeric(1), "value")
    errors <- vapply(integrals, `[[`, numeric(1), "error")
    ez <- values[2] / values[1]
    ez2 <- values[3] / values[1]
    stopifnot(values[1] > 0, ez2 > ez^2)
    list(log_marginal = mode$objective + log(values[1]), eap = x$sigma * ez,
         sd = x$sigma * sqrt(ez2 - ez^2), ez2 = ez2, mode_z = mode$maximum,
         scaled_integrals = values, scaled_numeric_errors = errors, log_scale = mode$objective)
  })
  log_mass <- vapply(per_person, `[[`, numeric(1), "log_marginal")
  # Each unweighted response likelihood is <= 1, giving absolute normal-tail bounds.
  tails <- c(mass = 2 * pnorm(-bound), first_absolute_theta_moment = x$sigma * 2 * dnorm(bound),
             second_theta_moment = x$sigma^2 * 2 * (bound * dnorm(bound) + pnorm(-bound)))
  list(nll = -sum(log_mass), eap = vapply(per_person, `[[`, numeric(1), "eap"),
       sd = vapply(per_person, `[[`, numeric(1), "sd"),
       log_sigma_nll_score = -sum(vapply(per_person, `[[`, numeric(1), "ez2") - 1),
       settings = settings, per_person = per_person, absolute_tail_bounds = tails,
       numeric_relative_mass_error_sum = sum(vapply(per_person, function(p)
         p$scaled_numeric_errors[1] / p$scaled_integrals[1], numeric(1))),
       tail_relative_mass_bound_sum = sum(exp(log(tails[1]) - log_mass)))
}

# Runnable analytic limiting check: every category has probability 1/5 as SD -> 0.
zero <- expand_parameters(c(rep(0, 23), log(1e-10)))
stopifnot(max(abs(exp(log_probabilities(zero, 0)) - 0.2)) < 1e-14)
for (settings in input$integration |> split(seq_len(nrow(input$integration)))) {
  control <- continuous_summary(zero, as.list(settings))
  stopifnot(abs(control$nll - 480 * log(5)) < 1e-8,
            max(abs(control$sd - 1e-10)) < 1e-18,
            abs(control$log_sigma_nll_score) < 1e-8)
}
for (name in names(cases)) {
  output <- file.path(args[3], paste0(name, "_r.json"))
  stopifnot(!file.exists(output))
  x <- expand_parameters(unlist(cases[[name]]$coordinates))
  result <- list(id = name, finite = lapply(rules, function(r) finite_summary(x, r)),
    probabilities = lapply(c(-3, 0, 3), function(t) exp(log_probabilities(x, t))[1:20, ]),
    continuous = lapply(seq_len(nrow(input$integration)), function(i)
      continuous_summary(x, as.list(input$integration[i, ]))),
    runtime = list(R = R.version.string, statmod = as.character(packageVersion("statmod")),
                   jsonlite = as.character(packageVersion("jsonlite"))))
  jsonlite::write_json(result, output, digits = NA, pretty = TRUE, auto_unbox = TRUE,
                       na = "null", null = "null")
  cat(name, "continuous NLL", result$continuous[[2]]$nll, "\n")
}
